File: HaNoi_2324/Bai5/DOANTOT.cpp
#include <iostream>
#include <vector>
#include <fstream>

using namespace std;

struct Segment {
    int l, r;
};

int main() {
    const int MOD = 1e9 + 7;
    ifstream infile("DOANTOT.INP");
    ofstream outfile("DOANTOT.OUT");

    int N;
    infile >> N;

    vector<Segment> segments(N);
    for (int i = 0; i < N; ++i) {
        infile >> segments[i].l >> segments[i].r;
    }

    vector<vector<Segment>> groups;
    vector<int> results;

    for (int i = 0; i < N; ++i) {
        Segment current_segment = segments[i];
        bool merged = false;

        // Kiểm tra xem đoạn thẳng hiện tại có giao nhau với nhóm nào không
        for (int j = 0; j < groups.size(); ++j) {
            vector<Segment> &group = groups[j];
            // Kiểm tra giao nhau
            for (const auto& seg : group) {
                if (seg.l <= current_segment.r && seg.r >= current_segment.l) {
                    group.push_back(current_segment);
                    merged = true;
                    break;
                }
            }
            if (merged) break;
        }

        if (!merged) {
            // Tạo nhóm mới
            groups.push_back({current_segment});
        }

        // Tính tích độ tốt
        long long total_goodness = 1;
        for (const auto& group : groups) {
            total_goodness = (total_goodness * group.size()) % MOD;
        }

        results.push_back(total_goodness);
    }

    // Ghi kết quả ra tệp
    for (const auto& res : results) {
        outfile << res << endl;
    }

    infile.close();
    outfile.close();
    return 0;
}

File: THCS/2023_2024/HaNoi_2324/Bai5/DOANTOT.py
class Segment:
    def __init__(self, l, r):
        self.l = l
        self.r = r

def main():
    MOD = 10**9 + 7
    
    # Đọc dữ liệu
    with open('DOANTOT.INP', 'r') as f:
        N = int(f.readline().strip())
        segments = []
        for _ in range(N):
            L, R = map(int, f.readline().strip().split())
            segments.append(Segment(L, R))
    
    # Danh sách lưu các nhóm đoạn thẳng
    groups = []
    results = []
    
    for i in range(N):
        current_segment = segments[i]
        merged = False
        
        # Kiểm tra xem đoạn thẳng hiện tại có giao nhau với nhóm nào không
        for j in range(len(groups)):
            group = groups[j]
            # Kiểm tra giao nhau
            if any(seg.l <= current_segment.r and seg.r >= current_segment.l for seg in group):
                group.append(current_segment)
                merged = True
                break
        
        if not merged:
            # Tạo nhóm mới
            groups.append([current_segment])
        
        # Tính tích độ tốt
        total_goodness = 1
        for group in groups:
            total_goodness = (total_goodness * len(group)) % MOD
        
        results.append(total_goodness)

    # Ghi kết quả ra tệp
    with open('DOANTOT.OUT', 'w') as f:
        for res in results:
            f.write(str(res) + '\n')


if __name__ == "__main__":
    main()
